### PR TITLE
istioctl: support `gateway` as input type

### PR DESCRIPTION
**Pull request by howardjohn**
This allows things like `istioctl proxy-config listeners gateway/my-gateway`. Also
included some bug fixes for namespace resolution (which is valid even
without the Gateway part).

**Please provide a description of this PR:**

### CODE REVIEW

**Review by hanxiaop — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
Seems no change to cmd to support `istioctl proxy-config listeners gateway/my-gateway`, do I miss something

**Comment by howardjohn** [comment]
InferPodInfoFromTypedResource is changed to support gateway and proxy-config uses that.

File: istioctl/pkg/util/handlers/handlers_test.go [hunk 1]
```diff
@@ -56,6 +56,42 @@ func TestInferPodInfo(t *testing.T) {
 			wantPodName:   "istio-security-post-install-1.2.2-bm9w2",
 			wantNamespace: "istio-system",
 		},
+		{
+			proxyName:     "service/istiod",
```
Comment (hzxuzhonghu) [inline istioctl/pkg/util/handlers/handlers_test.go:60]: Not sure about this usage and "gateway.gateway.networking.k8s.io/istiod"

Curentl it is `istioctl proxy-config clusters <pod-name[.namespace]> `, right

Comment (howardjohn) [reply istioctl/pkg/util/handlers/handlers_test.go:60]: No, current usage is `name.namespace` or `type/name.namespace`. 

The existing bug is type cannot have a `.` when it's perfectly valid syntax in k8s

Comment (hzxuzhonghu) [reply istioctl/pkg/util/handlers/handlers_test.go:60]: Oh i noticed it, `listener [<type>/]<name>[.<namespace>]` looks very tricky

File: istioctl/pkg/util/handlers/handlers_test.go [hunk 2]
```diff
@@ -56,6 +56,42 @@ func TestInferPodInfo(t *testing.T) {
...
+			namespace:     "namespace",
+			wantPodName:   "service/istiod",
+			wantNamespace: "namespace",
+		},
+		{
+			proxyName:     "service/istiod.istio-system",
+			namespace:     "namespace",
+			wantPodName:   "service/istiod",
```
Comment (hzxuzhonghu) [inline istioctl/pkg/util/handlers/handlers_test.go:74]: I am confused about the pod names here you added

Comment (howardjohn) [reply istioctl/pkg/util/handlers/handlers_test.go:74]: I think the problem is before this function returned "pod name" but was later changed to "resource name" without updating the tests. I can fix it but it's not really in scope of my PR.

Comment (hzxuzhonghu) [reply istioctl/pkg/util/handlers/handlers_test.go:74]: Can you also explain what does below cmd do with this pr?

`istioctl pc listener gateway.gateway.networking.k8s.io/istiod`

Comment (howardjohn) [reply istioctl/pkg/util/handlers/handlers_test.go:74]: The same thing `svc/xyz`, `deployment/xyz`, etc. Gets a pod associated with that resource. There is a direct mapping from Service to Pod, and similarly there is a direct mapping from Gateway to Pod which we use here.

This allows users to think about higher level construct (Gateway) and not worry about implementation detail (pod name)

Comment (hzxuzhonghu) [reply istioctl/pkg/util/handlers/handlers_test.go:74]: IC, that is to say i can use istio gateway as well, though I do not think anyone will use it like this.

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**